lib/active_model/serializer/associations.rb

```diff
@@ -73,12 +76,19 @@ def has_one(name, options = {})
         #
         # @api private
         #
-        def associate(reflection)
+        def associate(reflection, block)
           self._reflections = _reflections.dup
 
-          define_method reflection.name do
-            object.send reflection.name
-          end unless method_defined?(reflection.name)
+          reflection_name = reflection.name
+          if block
+            serialized_associations[reflection_name] = ->(instance) { instance.instance_eval(&block) }
+          else
+            serialized_associations[reflection_name] = ->(instance) { instance.object.send(reflection_name) }
+          end
+
+          define_method reflection_name do
```

Comment:
Yeah that's what was holding me back, because even after fixing the tests, there are caching issues (I believe caching is broken tbh).